thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/alert/filter/DimensionsRecipientAlertFilter.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.pinot.thirdeye.detection.alert.StatefulDetectionAlertFilter;
+import org.apache.pinot.thirdeye.detection.annotation.AlertFilter;
+
+
+/**
+ * The detection alert filter that can send notifications through multiple channels
+ * to a set of unconditional and another set of conditional recipients, based on the
+ * value of a specified anomaly dimension combinations.
+ *
+ * <pre>
+ * dimensionRecipients:
+ *   - dimensions:
+ *       country: IN
+ *       device: Android
+ *     notify:
+ *       jiraScheme:
+ *         project: ANDROID
+ *         assignee: android-oncall
+ *       emailScheme:
+ *         recipients:
+ *           to:
+ *           - "android-team@comany.com"
+ *   - dimension:
+ *       country: US
+ *       device: IOS
+ *     notify:
+ *       jiraScheme:
+ *         project: IOS
+ *         assignee: ios-oncall
+ * </pre>
+ */
+@AlertFilter(type = "DIMENSIONS_ALERTER_PIPELINE")
+public class DimensionsRecipientAlertFilter extends StatefulDetectionAlertFilter {
+  protected static final String PROP_DETECTION_CONFIG_IDS = "detectionConfigIds";
+  protected static final String PROP_DIMENSION = "dimensions";
+  protected static final String PROP_NOTIFY = "notify";
+  protected static final String PROP_DIMENSION_RECIPIENTS = "dimensionRecipients";
+  private static final String PROP_SEND_ONCE = "sendOnce";
+
+  private Map<String, Object> defaultNotificationSchemeProps = new HashMap<>();
+  final List<Map<String, Object>> dimensionRecipients;
+  final List<Long> detectionConfigIds;
+  final boolean sendOnce;
+
+  public DimensionsRecipientAlertFilter(DataProvider provider, DetectionAlertConfigDTO config, long endTime) {
+    super(provider, config, endTime);
+    for (Map.Entry<String, Map<String, Object>> schemeProps : this.config.getAlertSchemes().entrySet()) {
+      defaultNotificationSchemeProps.put(schemeProps.getKey(), new HashMap<>(schemeProps.getValue()));
+    }
+    this.dimensionRecipients = ConfigUtils.getList(this.config.getProperties().get(PROP_DIMENSION_RECIPIENTS));
+    this.detectionConfigIds = ConfigUtils.getLongs(this.config.getProperties().get(PROP_DETECTION_CONFIG_IDS));
+    this.sendOnce = MapUtils.getBoolean(this.config.getProperties(), PROP_SEND_ONCE, true);
+  }
+
+  @Override
+  public DetectionAlertFilterResult run(Map<Long, Long> vectorClocks, long highWaterMark) {
+    DetectionAlertFilterResult result = new DetectionAlertFilterResult();
+    final long minId = getMinId(highWaterMark);
+
+    Set<MergedAnomalyResultDTO> anomalies = this.filter(this.makeVectorClocks(this.detectionConfigIds), minId);
+
+    for (Map<String, Object> dimensionRecipient : this.dimensionRecipients) {
+      Map<String, String> dimensionFilters = ConfigUtils.getMap(dimensionRecipient.get(PROP_DIMENSION));
+      Set<MergedAnomalyResultDTO> notifyAnomalies = new HashSet<>();
+      for (MergedAnomalyResultDTO anomaly : anomalies) {
+        if (anomaly.getDimensions().contains(dimensionFilters)) {
```

Comment:
Actually, we've deprecated the `DimensionMap` class, because it's a `Map<String, String>` instead of a MultiMap. The `anomaly.getDimensions()` field is not fully populated in some cases. Can use `MetricEntity.fromUrn(anomaly.getMetricUrn()).getFilters()` to avoid the bugs.

Reply:
Updated